src/System.Net.Security/src/System/Net/NegotiateStreamPal.Unix.cs

```diff
@@ -0,0 +1,298 @@
+                return (string.IsNullOrEmpty(credential.UserName) || string.IsNullOrEmpty(credential.Password)) ?
+                    new SafeFreeNegoCredentials(string.Empty, string.Empty, string.Empty) :
+                    new SafeFreeNegoCredentials(credential.UserName, credential.Password, credential.Domain);
+            }
+            catch(Exception ex)
+            {
+                throw new Win32Exception(NTE_FAIL, ex.Message);
+            }
+        }
+
+        internal static SecurityStatusPal InitializeSecurityContext(
+            SafeFreeCredentials credentialsHandle,
+            ref SafeDeleteContext securityContext,
+            string spn,
+            ContextFlagsPal requestedContextFlags,
+            SecurityBuffer[] inSecurityBufferArray,
+            SecurityBuffer outSecurityBuffer,
+            ref ContextFlagsPal contextFlags)
+        {
+            // TODO (Issue #3718): The second buffer can contain a channel binding which is not supported
+            if ((null != inSecurityBufferArray) && (inSecurityBufferArray.Length > 1))
+            {
+                throw new PlatformNotSupportedException(SR.net_nego_channel_binding_not_supported);
+            }
+
+            if ((null != inSecurityBufferArray) && (inSecurityBufferArray.Length > 0) && (inSecurityBufferArray[0].type == SecurityBufferType.ChannelBindings))
+            {
+                throw new PlatformNotSupportedException(SR.net_nego_channel_binding_not_supported);
+            }
+
+            return EstablishSecurityContext(
+                (SafeFreeNegoCredentials)credentialsHandle,
+                ref securityContext,
+                false,
+                spn,
+                requestedContextFlags,
+                ((inSecurityBufferArray != null && inSecurityBufferArray.Length != 0) ? inSecurityBufferArray[0] : null),
+                outSecurityBuffer,
+                ref contextFlags);
+        }
+
+        internal static SecurityStatusPal CompleteAuthToken(
+            ref SafeDeleteContext securityContext,
+            SecurityBuffer[] inSecurityBufferArray)
+        {
+            return new SecurityStatusPal(SecurityStatusPalErrorCode.OK);
+        }
+
+        internal static SecurityStatusPal AcceptSecurityContext(
+            SafeFreeCredentials credentialsHandle,
+            ref SafeDeleteContext securityContext,
+            ContextFlagsPal requestedContextFlags,
+            SecurityBuffer[] inSecurityBufferArray,
+            SecurityBuffer outSecurityBuffer,
+            ref ContextFlagsPal contextFlags)
+        {
+            throw new PlatformNotSupportedException();
+        }
+
+        internal static void ValidateImpersonationLevel(TokenImpersonationLevel impersonationLevel)
+        {
+            if (impersonationLevel != TokenImpersonationLevel.Identification)
+            {
+                throw new ArgumentOutOfRangeException("impersonationLevel", impersonationLevel.ToString(),
```

Comment:
There's actually a minor perf advantage to calling ToString explicitly (though we're about to throw, so that's unlikely to make a material difference).  By calling ToString, we only incur the string allocation, whereas if the value type impersonationLevel is passed, it first needs to be boxed, and then ToString will be called on that.